asset/background.js

```diff
@@ -0,0 +1,104 @@
+
+  KeyQ: { key: { plain: 'q', shifted: 'Q' }, code: 'KeyQ' },
+  KeyW: { key: { plain: 'w', shifted: 'W' }, code: 'KeyW' },
+  KeyE: { key: { plain: 'j', shifted: 'J' }, code: 'KeyE' },
+  KeyR: { key: { plain: 'f', shifted: 'F' }, code: 'KeyR' },
+  KeyT: { key: { plain: 'g', shifted: 'G' }, code: 'KeyT' },
+  KeyY: { key: { plain: 'y', shifted: 'Y' }, code: 'KeyY' },
+  KeyU: { key: { plain: 'p', shifted: 'P' }, code: 'KeyU' },
+  KeyI: { key: { plain: 'u', shifted: 'U' }, code: 'KeyI' },
+  KeyO: { key: { plain: 'l', shifted: 'L' }, code: 'KeyO' },
+  KeyP: { key: { plain: ';', shifted: ':' }, code: 'KeyP' },
+  BracketLeft: { key: { plain: '[', shifted: '{' }, code: 'BracketLeft' },
+  BracketRight: { key: { plain: ']', shifted: '}' }, code: 'BracketRight' },
+  KeyA: { key: { plain: 'a', shifted: 'A' }, code: 'KeyA' },
+  KeyS: { key: { plain: 's', shifted: 'S' }, code: 'KeyS' },
+  KeyD: { key: { plain: 'e', shifted: 'E' }, code: 'KeyD' },
+  KeyF: { key: { plain: 't', shifted: 'T' }, code: 'KeyF' },
+  KeyG: { key: { plain: 'd', shifted: 'D' }, code: 'KeyG' },
+  KeyH: { key: { plain: 'h', shifted: 'H' }, code: 'KeyH' },
+  KeyJ: { key: { plain: 'n', shifted: 'N' }, code: 'KeyJ' },
+  KeyK: { key: { plain: 'i', shifted: 'I' }, code: 'KeyK' },
+  KeyL: { key: { plain: 'o', shifted: 'O' }, code: 'KeyL' },
+  Semicolon: { key: { plain: 'r', shifted: 'R' }, code: 'Semicolon' },
+  Quote: { key: { plain: "'", shifted: '"' }, code: 'Quote' },
+  KeyZ: { key: { plain: 'z', shifted: 'Z' }, code: 'KeyZ' },
+  KeyX: { key: { plain: 'x', shifted: 'X' }, code: 'KeyX' },
+  KeyC: { key: { plain: 'c', shifted: 'C' }, code: 'KeyC' },
+  KeyV: { key: { plain: 'v', shifted: 'V' }, code: 'KeyV' },
+  KeyB: { key: { plain: 'b', shifted: 'B' }, code: 'KeyB' },
+  KeyN: { key: { plain: 'k', shifted: 'K' }, code: 'KeyN' },
+  KeyM: { key: { plain: 'm', shifted: 'M' }, code: 'KeyM' },
+  Comma: { key: { plain: ',', shifted: '<' }, code: 'Comma' },
+  Period: { key: { plain: '.', shifted: '>' }, code: 'Period' },
+  Slash: { key: { plain: '/', shifted: '?' }, code: 'Slash' }
+};
+
+chrome.input.ime.onFocus.addListener(function(context) {
+  contextID = context.contextID;
+});
+
+function updateShiftState(keyData) {
+  var isShifted = (keyData.shiftKey && !keyData.capsLock) || (!keyData.shiftKey && keyData.capsLock);
+  shiftState = isShifted ? Shift.SHIFTED : Shift.PLAIN;
+}
+
+function isPureModifier(keyData) {
+  return keyData.key == 'Shift' || keyData.key == 'Ctrl' || keyData.key == 'Alt';
+}
+
+chrome.input.ime.onKeyEvent.addListener(function(engineID, keyData) {
+  var handled = false;
+
+  updateShiftState(keyData);
+
+  if (assetKeymap[keyData.code]) {
+    // If these arn't relevent keypresses, then quit early.
+    if (isPureModifier(keyData) || keyData.type !== 'keydown') {
+      return handled;
+    }
+
+    // create a new object. Better than mutating keyData directly.
+    var remappedKeyData = keyData;
```

Comment:
You only really need to the string value to supply to commitText later. So you can simplify this to: var emit = assetKeymap[keyData.code].key[shiftState] and then in commitText you pass the emit string.

Reply:
Makes sense. Updated ^_^